aws-api-appsync/src/main/java/com/amplifyframework/datastore/appsync/SerializedModelAdapter.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.datastore.appsync;
+
+import com.amplifyframework.core.model.ModelField;
+import com.amplifyframework.core.model.ModelSchema;
+import com.amplifyframework.util.GsonObjectConverter;
+
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Deserializer for SerializedModel. Helpful to deserialize from the graphql response.
+ */
+public final class SerializedModelAdapter
+        implements JsonDeserializer<SerializedModel>, JsonSerializer<SerializedModel> {
+    private SerializedModelAdapter() {}
+
+    /**
+     * Registers an adapter with a Gson builder.
+     * @param builder A gson builder
+     */
+    public static void register(GsonBuilder builder) {
+        builder.registerTypeAdapter(SerializedModel.class, new SerializedModelAdapter());
+    }
+
+    @Override
+    public JsonElement serialize(SerializedModel src, Type typeOfSrc, JsonSerializationContext context) {
+        ModelSchema schema = src.getModelSchema();
+
+        JsonObject result = new JsonObject();
+        result.add("id", context.serialize(src.getId()));
+        result.add("modelSchema", context.serialize(schema));
```

Comment:
The `ModelSchema` is only ever used locally on the client, as metadata. It never gets saved to the SQL tables, and it never gets passed to AppSync.

At various points in the code, I need to rectify the shape of the `serializedData` -- iterating over its fields, to understand the type of each one. There are two ways to achieve this:
1. Just store the `modelName`, and then retrieve the `ModelSchema` for that `modelName`, from the `ModelSchemaRegistry`;
2. Store the entire `ModelSchema` next to the `SerializedModel`. This is a little wasteful, but it's pretty convenient, and it only ever happens in memory. Whenever I need to understand the schema for the data, it's right next to it.

I don't think, strictly speaking, there is a need to serialize the `ModelSchema`. But, it happens to be convenient (at least) in test code. Since I had a deserialize, it felt ~right to implement the serialize method at the same time, in the same place, to show the symmetry of the operation.

Reply:
Thanks for clarifying!   Since it isn't getting saved in SQLite, this is not really a concern any more!